### PR TITLE
feat(aapd-253): dynamic form file upload

### PR DESCRIPTION
**Pull request by bryn-pins**
## Ticket Number

https://pins-ds.atlassian.net/browse/aapd-253
https://pins-ds.atlassian.net/browse/aapd-255

## Checklist

- [ ] Requires infrastructure changes
- [ ] If adding or remove environment variables (e.g. in `docker-compose.yaml`) then I have updated the appropriate Helm chart
- [ ] I have updated the documentation accordingly
- [x] My commit history in this PR is linear
- [x] New features have tests
- [ ] Breaking change (team conversation required)

## Important

Please do not merge from `main` (please only [rebase](https://github.com/foundry4/appeal-planning-decision/wiki/An-intro-to-Git-Rebase)). This keeps the history linear and easier to debug.

### CODE REVIEW

File: packages/forms-web-app/src/dynamic-forms/has-questionnaire/questions.js [hunk 1]
```diff
@@ -79,7 +79,7 @@ exports.questions = {
 		title: 'Who did you notify?',
 		question: 'Who did you notify about this application?',
 		fieldName: 'notified-who',
-		validators: [new RequiredValidator('You must add your documents')]
```
Comment (bryn-pins) [inline packages/forms-web-app/src/dynamic-forms/has-questionnaire/questions.js:82]: removed this for now until validators decided

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js [hunk 2]
```diff
@@ -10,6 +32,146 @@ class MultiFileUploadQuestion extends Question {
...
+			// 		href: '#'
+			// 	}));
+			// 	return this.#showError(res, viewModel);
+			// }
+
+			// map the response back to journeyResponse
+			journeyResponse.answers[this.fieldName] = responseToSave.answers[this.fieldName];
+
+			// update journey based on response
+			const updatedJourney = new journey.constructor(journeyResponse);
```
Comment (bryn-pins) [inline packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js:120]: wasn't sure on the best way to achieve this, alternative could be to have an update journey function

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js [hunk 3]
```diff
@@ -10,6 +32,146 @@ class MultiFileUploadQuestion extends Question {
...
+				);
+			}
+
+			return res.redirect(
+				updatedJourney.getNextQuestionUrl(sectionObj.segment, this.fieldName, false)
+			);
+		} catch (err) {
+			logger.error(err);
+
+			return this.#showError(res, journey, sectionObj, {
```
Comment (chrishawldev) [inline packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js:136]: I'm guessing this is for a scenario where validation ext passes but there is some other issue that prevents the files being saved?

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js [hunk 4]
```diff
@@ -10,6 +32,146 @@ class MultiFileUploadQuestion extends Question {
...
+				const mappedDoc = mapMultiFileDocumentToSavedDocument(document, document?.name, file.name);
+
+				result.push(mappedDoc);
+			}
+		}
+
+		return result;
+	}
+
+	#showError(res, journey, sectionObj, customViewData) {
```
Comment (chrishawldev) [inline packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js:168]: Would this work at a level above? As in should we implement showError on the base question? Perhaps we should be doing this in every scenario where we save data (e.g when cosmos fails to save an answer)?

Comment (chrishawldev) [reply packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js]: Ah I can see you're not handling this with the render page method and passing the error summary in there in the catch block, looks good 👍

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js (hunk 3 above)
Comment (bryn-pins) [reply packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js]: Potentially yes, have been unable to trigger it, but can leave in for now

Comment (chrishawldev) [reply packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js]: I think it's a good thing to get a test against. Could we trigger it in a unit test by mocking out one of the downstream dependencies to throw an error?

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js (hunk 2 above)
Comment (chrishawldev) [reply packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js:136]: I think thats a good thing to keep in mind if we need to do this anywhere else in the future, but if this behaviour is just a one off then this should be fine.

**Comment by bryn-pins** [comment]
@chrishawldev I can't seem to comment on the unresolved thread.
That was for the upload-and-remain-on-page bit I left a comment on not knowing what it was for right?
Was there anything left to do on this?

File: packages/common/src/document-types.js [hunk 5]
```diff
@@ -221,6 +221,14 @@ const documentTypes = {
 		involvement: '',
 		horizonDocumentType: 'Appellant Final Comments',
 		horizonDocumentGroupType: 'Evidence'
+	},
+	dynamic: {
```
Comment (chrishawldev) [inline packages/common/src/document-types.js:225]: What's the use case for the dynamic document type?

**Review by chrishawldev — APPROVED** [review]

**Review by chrishawldev — APPROVED** [review]

File: packages/common/src/document-types.js (hunk 5 above)
Comment (bryn-pins) [reply packages/common/src/document-types.js:225]: So the document api requires meta data to be added alongside the file upload which comes from these objects.

It also runs integration style tests against each of the document types from this file to ensure they can be uploaded and fetched again.

So a large part of this is so I didn't need to rewrite too much of the existing document upload logic/tests/potentially break the existing appeals journey

I was under the impression we would want to continue using this, but would be adding new document types for each upload question, at which point we have a single source of every document type we have.

However I wasn't sure what the meta data needed for the various uploads was going to be so I created a generic one as a placeholder (the tickets for uploads don't specify any metadata needed at the moment)

So:
Maybe we don't need and we need to refactor the existing document upload/tests
Maybe we can just leave as this as is, as a catch all for docs that don't need this metadata
Maybe we add a new doc type for each file upload question

**Pull request merged**